File: features/usageApi.ts
import {apiRequest, deviceTimezone} from './apiClient';

export type QuotaBucket = {
    used: number;
    limit: number;
    remaining: number;
};

export type UsageSnapshot = {
    date: string;
    timezone: string;
    hasPremium: boolean;
    dailyCards: QuotaBucket;
    adUnlocks: QuotaBucket;
    interpretations: QuotaBucket;
};

export type AdSession = {
    adToken: string;
    minWatchSeconds: number;
    date: string;
};

const withTz = (path: string): string => {
    const tz = encodeURIComponent(deviceTimezone());
    return path.includes('?') ? `${path}&tz=${tz}` : `${path}?tz=${tz}`;
};

export const fetchUsage = () => apiRequest<UsageSnapshot>(withTz('/api/v1/usage'));

export const startAdSession = () =>
    apiRequest<AdSession>(withTz('/api/v1/usage/ad-session'), {method: 'POST', body: {}});

export const consumeDailyCard = (adToken?: string) =>
    apiRequest<UsageSnapshot>(withTz('/api/v1/usage/daily-card'), {
        method: 'POST',
        body: adToken ? {adToken} : {},
    });

File: ios-swift/MoonCompass/Core/Networking/Endpoints/UsageAPI.swift
import Foundation

struct QuotaBucket: Decodable, Sendable, Equatable {
    var used: Int
    var limit: Int
    var remaining: Int
}

/// Today's quotas as the server counts them, in the device's time zone
/// (the client sends `X-Timezone` on every request).
struct UsageSnapshot: Decodable, Sendable, Equatable {
    var date: String
    var timezone: String
    var hasPremium: Bool
    var dailyCards: QuotaBucket
    var interpretations: QuotaBucket
}

private struct EmptyBody: Encodable, Sendable {}

struct UsageAPI: Sendable {
    let client: APIClient

    func usage() async throws -> UsageSnapshot {
        try await client.send(.get, "/api/v1/usage", auth: .required)
    }

    /// Spends one daily-card draw. Throws `quota_exceeded` (429) at the limit.
    ///
    /// The ad-unlock flow of the React Native client is not ported: its "ad"
    /// was the app's own splash video, which App Review could fairly call
    /// misleading, so the iOS client never sends an ad token.
    func consumeDailyCard() async throws -> UsageSnapshot {
        try await client.send(.post, "/api/v1/usage/daily-card", body: EmptyBody(), auth: .required)
    }
}
